20240927-multiple-object-deletion.md

```diff
@@ -0,0 +1,163 @@
+
+For further context, the gateway currently supports Amazon S3's [DeleteObjects](https://docs.aws.amazon.com/AmazonS3/latest/API/API_DeleteObjects.html) operation, which allows for the deletion of multiple objects with a single HTTP request. However, the lack of native support for bulk deletions in the metainfo RPC service makes this operation subject to the aforementioned issues, as the gateway must translate every S3 request to a series of RPC requests.
+
+### Goals
+
+The goal of the implementation described by this document is to allow multiple objects to be deleted in a single request to the satellite.
+
+### Approach / Design
+
+#### Satellite
+
+The satellite metainfo RPC service must be extended with an endpoint allowing for the deletion of multiple objects in a single request.
+
+Requests directed to this endpoint must contain the following information:
+
+- The requester's API key
+- The bucket from which objects will be deleted
+- An indication of whether governance mode Object Lock restrictions should be bypassed
+- An indication of whether quiet mode should be enabled.
+  - With quiet mode enabled, only errors are returned; information about successful deletions is omitted from responses. Supporting this mode would allow us to save bandwidth on clients that use it.
+- A list of objects to delete
+
+The endpoint must respond with the following information:
+
+- A list containing the details of each successful deletion, including:
+  - The deleted object's key
+  - The deleted object's version (if the version wasn't specified in the request)
+  - The version ID of the delete marker (if applicable)
+- A list containing the details of each failed deletion, including:
+  - An error code
+  - The object's key
+  - The object's version ID (if it was provided in the request)
+
+#### Libuplink
+
+Libuplink must provide a method through which the satellite metainfo RPC service's multiple object deletion endpoint is contacted. This is necessary for the gateway to function.
+
+#### Gateway
+
+The gateway must be given a configuration option that indicates whether the new multiple object deletion feature should be used. When it is enabled, it must invoke the new Libuplink method.
+
+#### Protobuf
+
+The protobuf definition for the metainfo RPC service must include a method for deleting multiple objects. Additionally, request and response types must be defined.
+
+```proto
+service Metainfo {
+    rpc DeleteObjects(DeleteObjectsRequest) returns (DeleteObjectsResponse);
+}
+
+message DeleteObjectsRequest {
+    RequestHeader header = 15;
+
+    bytes bucket = 1;
+    bool bypass_governance_retention = 2;
+    bool quiet = 3;
+    repeated DeleteObjectsRequestItem items = 4;
+}
+
+message DeleteObjectsRequestItem {
+    bytes encrypted_object_key = 1;
+    bytes object_version = 2;
+}
+
+message DeleteObjectsResponse {
+	repeated DeleteObjectsResponseDeletedItem deleted = 1;
+	repeated DeleteObjectsResponseErrorItem errors = 2;
```

Comment:
nit: spaces instead of tabs for consistency with rest of the proto example:

```suggestion
    repeated DeleteObjectsResponseErrorItem errors = 2;
```